### PR TITLE
[Woo POS] Refactor event translation to alerts and messages

### PR DESCRIPTION
**Pull request by joshheald**
Part of: #12868 

## Description

This PR refactors our translation from `CardPresentPaymentEvent` to the view models that we're presenting in the POS.

Apologies that it's so long. I could split it into connection and payment alerts if that would help, but it would leave one of them broken between the PRs, if I did.

While we're writing the `CardPresentPaymentService` initially for Point of Sale, it's not bound to that product. In future, we may want other parts of the app to use this service, and the events might be presented differently from these other areas.

In this PR I've tried to make the boundary clearer with naming and enums. Anything with a `PointOfSale` prefix is specific to the POS implementation, but the `CardPresentPaymentEvent` and its associated details object aren't.

Additionally, this PR moves us closer to a single `show` event – the next enum case to incorporate to that is the `showFoundReaders` case, which could definitely be handled in the same way as the rest of the connection event details now.

## Screenshots

---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/Classes/POS/Presentation/Card Present Payments/PointOfSaleCardPresentPaymentEventPresentationStyle.swift [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+            return .alert(.updatingReader(viewModel: CardPresentPaymentUpdatingReaderAlertViewModel()))
+
+        case .updateFailed(let tryAgain, let cancelUpdate):
+            return .alert(.updateFailed(viewModel: CardPresentPaymentReaderUpdateFailedAlertViewModel()))
+
+        case .updateFailedNonRetryable(let cancelUpdate):
+            return .alert(.updateFailed(viewModel: CardPresentPaymentReaderUpdateFailedAlertViewModel()))
+
+        case .updateFailedLowBattery(let batteryLevel, let cancelUpdate):
+            return .alert(.updateFailed(viewModel: CardPresentPaymentReaderUpdateFailedAlertViewModel()))
```
Comment (jaclync) [inline WooCommerce/Classes/POS/Presentation/Card Present Payments/PointOfSaleCardPresentPaymentEventPresentationStyle.swift:81]: non-blocking: might we want to mention the low battery info in the alert? Feel free to ignore this if you already have this as a future task.

File: WooCommerce/Classes/POS/Presentation/Card Present Payments/PointOfSaleCardPresentPaymentEventPresentationStyle.swift [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        case .preparingForPayment(cancelPayment: let cancelPayment):
+            return .message(.preparingForPayment)
+
+        case .tapSwipeOrInsertCard(inputMethods: let inputMethods, cancelPayment: let cancelPayment):
+            return .message(.tapSwipeOrInsertCard)
+
+        case .success(done: let done):
+            return .message(.success)
+
+        case .error(error: let error, tryAgain: let tryAgain, cancelPayment: let cancelPayment):
```
Comment (jaclync) [inline WooCommerce/Classes/POS/Presentation/Card Present Payments/PointOfSaleCardPresentPaymentEventPresentationStyle.swift:93]: nit: this case seems to be emitted in `CardReaderTransactionAlertsProviding` currently, but the name is pretty generic and it's unclear whether it's from the connection flow or the payment flow in which we show the error UI differently. Maybe this case can be renamed to something like `transactionError`/`paymentError` to provide the context of the error.

File: WooCommerce/Classes/POS/Presentation/Card Present Payments/PointOfSaleCardPresentPaymentEventPresentationStyle.swift [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        case .processing:
+            return .message(.processing)
+
+        case .displayReaderMessage(message: let message):
+            return .message(.displayReaderMessage(message: message))
+
+        case .cancelledOnReader:
+            return .message(.cancelledOnReader)
+
+        /// Not-yet supported types
```
Comment (jaclync) [inline WooCommerce/Classes/POS/Presentation/Card Present Payments/PointOfSaleCardPresentPaymentEventPresentationStyle.swift:108]: just a clarifying question, do we plan to support these in M1?

File: WooCommerce/Classes/POS/Presentation/Card Present Payments/PointOfSaleCardPresentPaymentEventPresentationStyle.swift [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        case .tapSwipeOrInsertCard(inputMethods: let inputMethods, cancelPayment: let cancelPayment):
+            return .message(.tapSwipeOrInsertCard)
+
+        case .success(done: let done):
+            return .message(.success)
+
+        case .error(error: let error, tryAgain: let tryAgain, cancelPayment: let cancelPayment):
+            return .message(.error)
+
+        case .errorNonRetryable(error: let error, cancelPayment: let cancelPayment):
```
Comment (jaclync) [inline WooCommerce/Classes/POS/Presentation/Card Present Payments/PointOfSaleCardPresentPaymentEventPresentationStyle.swift:96]: nit: similar to the `error` case above, maybe this case name could provide more context.

File: WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift [hunk 5]
```diff
@@ -141,27 +141,35 @@ private extension PointOfSaleDashboardViewModel {
...
-                return alertDetails
+                return alertType
             }
             .assign(to: &$cardPresentPaymentAlertViewModel)
         cardPresentPaymentService.paymentEventPublisher
-            .map { event in
-                guard case let .showPaymentMessage(messageType) = event else {
+            .map { event -> PointOfSaleCardPresentPaymentMessageType? in
+                guard case let .show(eventDetails) = event,
+                      case let .message(messageType) = eventDetails.pointOfSalePresentationStyle else {
```
Comment (jaclync) [inline WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift:155]: Nice mapping from one event to two presentation types!

File: WooCommerce/Classes/POS/Presentation/PointOfSaleCardPresentPaymentAlert.swift [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        case .scanningFailed(let alertViewModel):
+            PointOfSaleCardPresentPaymentScanningForReadersFailedView(viewModel: alertViewModel)
+        case .bluetoothRequired(let alertViewModel):
+            PointOfSaleCardPresentPaymentBluetoothRequiredAlertView(viewModel: alertViewModel)
+        case .foundReader(let alertViewModel):
+            PointOfSaleCardPresentPaymentFoundReadersView(viewModel: alertViewModel)
+        case .updatingReader(let alertViewModel):
+            CardPresentPaymentReaderUpdateInProgressView(viewModel: alertViewModel)
+        case .updateFailed(let alertViewModel):
+            CardPresentPaymentReaderUpdateFailedView(viewModel: alertViewModel)
```
Comment (jaclync) [inline WooCommerce/Classes/POS/Presentation/PointOfSaleCardPresentPaymentAlert.swift:23]: nit: any reasons these two views aren't renamed to `PointOfSale*` like the views in other cases?

**Review by jaclync — APPROVED** [review]
Thanks for taking on this refactoring, love the clear separation of POS and card-present payment events that can be used for all card-present payment use cases in the future including the existing app. I tested the happy path and some error case, and all worked nicely as before :shipit:

File: WooCommerce/Classes/POS/Presentation/Card Present Payments/PointOfSaleCardPresentPaymentEventPresentationStyle.swift (hunk 1 above)
Comment (joshheald) [reply WooCommerce/Classes/POS/Presentation/Card Present Payments/PointOfSaleCardPresentPaymentEventPresentationStyle.swift:81]: Yep, I'll tackle all the update details in a later task, they seemed lower priority than getting the (normal) connection flow working and this PR is big enough!

File: WooCommerce/Classes/POS/Presentation/Card Present Payments/PointOfSaleCardPresentPaymentEventPresentationStyle.swift (hunk 2 above)
Comment (joshheald) [reply WooCommerce/Classes/POS/Presentation/Card Present Payments/PointOfSaleCardPresentPaymentEventPresentationStyle.swift:93]: Good point, thanks!

File: WooCommerce/Classes/POS/Presentation/PointOfSaleCardPresentPaymentAlert.swift (hunk 6 above)
Comment (joshheald) [reply WooCommerce/Classes/POS/Presentation/PointOfSaleCardPresentPaymentAlert.swift:23]: Just that I've not done the Update views in this PR 😊 I'll rename them when I implement them though.

File: WooCommerce/Classes/POS/Presentation/Card Present Payments/PointOfSaleCardPresentPaymentEventPresentationStyle.swift (hunk 3 above)
Comment (joshheald) [reply WooCommerce/Classes/POS/Presentation/Card Present Payments/PointOfSaleCardPresentPaymentEventPresentationStyle.swift:108]: We might need to do `validatingOrder` – it's the first step after the reader connects though, so it's probably a simple implementation where we do exactly the same as `preparingForPayment`.

`selectSearchType` is about choosing between Tap to Pay and a Bluetooth reader, during the connection flow, so we don't need to do that in the MVP at all. We don't actually use it in the main app either, it's leftover from an old design.

File: WooCommerce/Classes/POS/Presentation/Card Present Payments/PointOfSaleCardPresentPaymentEventPresentationStyle.swift (hunk 2 above)
Comment (joshheald) [reply WooCommerce/Classes/POS/Presentation/Card Present Payments/PointOfSaleCardPresentPaymentEventPresentationStyle.swift:93]: Done in 2398b5e

**Pull request merged**